.github/workflows/bsim-tests.yaml

```diff
@@ -72,20 +75,72 @@ jobs:
           west forall -c 'git reset --hard HEAD'
 
+      - name: Check if Bluethooth files changed
+        uses: tj-actions/changed-files@v35
+        id: check-buetooth-files
+        with:
+          files: |
+            .github/workflows/bsim-tests.yaml
+            .github/workflows/bsim-tests-publish.yaml
+            west.yml
+            subsys/**
+            tests/bsim/bluetooth/**
+            samples/bluetooth/**
+            boards/posix/**
+            soc/posix/**
+            arch/posix/**
+
+      - name: Check if Networking files changed
+        uses: tj-actions/changed-files@v35
+        id: check-networking-files
+        with:
+          files: |
+            .github/workflows/bsim-tests.yaml
+            .github/workflows/bsim-tests-publish.yaml
+            west.yml
+            boards/posix/**
+            soc/posix/**
+            arch/posix/**
+            tests/bsim/**
+            samples/net/sockets/echo_*/**
+            modules/openthread/**
+            subsys/net/l2/openthread/**
+            include/zephyr/net/openthread.h
+            drivers/ieee802154/**
+            include/zephyr/net/ieee802154*
+
+      - name: Run Bluetooth Tests with BSIM
+        if: steps.check-buetooth-files.outputs.any_changed == 'true'
+        run: |
+          export ZEPHYR_BASE=${PWD}
+          WORK_DIR=${ZEPHYR_BASE}/bsim_bluetooth tests/bsim/bluetooth/compile.sh
+          RESULTS_FILE=${ZEPHYR_BASE}/${bsim_bluetooth_test_results_file} \
+          SEARCH_PATH=tests/bsim/bluetooth/ tests/bsim/run_parallel.sh
+
       - name: Run Networking Tests with BSIM
+        if: steps.check-networking-files.outputs.any_changed == 'true'
         run: |
           export ZEPHYR_BASE=${PWD}
-          WORK_DIR=${ZEPHYR_BASE}/bsim_out tests/bsim/net/compile.sh
-          RESULTS_FILE=${ZEPHYR_BASE}/${bsim_test_results_file} \
+          WORK_DIR=${ZEPHYR_BASE}/bsim_net tests/bsim/net/compile.sh
+          RESULTS_FILE=${ZEPHYR_BASE}/${bsim_networking_test_results_file} \
           SEARCH_PATH=tests/bsim/net/ tests/bsim/run_parallel.sh
 
-      - name: Upload Test Results
-        if: always()
+      - name: Upload Test Bluetooth Results
+        if: steps.check-buetooth-files.outputs.any_changed == 'true'
+        uses: actions/upload-artifact@v3
+        with:
+          name: bsim-bluetooth-test-results
+          path: |
+            ./bsim_bluetooth/bsim_results.xml
+            ${{ github.event_path }}
+
+      - name: Upload Test Networking Results
+        if: steps.check-networking-files.outputs.any_changed == 'true'
         uses: actions/upload-artifact@v3
         with:
-          name: bsim-test-results
+          name: bsim-networking-test-results
           path: |
-            ./bsim_out/bsim_results.xml
+            ./bsim_net/bsim_results.xml
             ${{ github.event_path }}
```

Comment:
Here you are dropping `if: always()`, which is there to upload the test result file even if one of the tests failed.

Since the `actions/upload-artifact` action does not treat missing files as a failure (it only outputs a warning by default), you can keep `if: always()` and merge these two steps into one:

```yaml
      - name: Upload Test Results
        if: always()
        uses: actions/upload-artifact@v3
        with:
          name: bsim-test-results
          path: |
            ./bsim_bluetooth/bsim_results.xml
            ./bsim_net/bsim_results.xml
            ${{ github.event_path }}
          if-no-files-found: warn
```

Reply:
I have applied the changes thanks for hint.